hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+                querySingleValue("SELECT JSON_QUERY(this, '$' EMPTY ARRAY ON EMPTY) AS c1 FROM test WHERE __key = 1"));
+        assertEquals(new HazelcastJsonValue("{}"),
+                querySingleValue("SELECT JSON_QUERY(this, '$' EMPTY OBJECT ON EMPTY) AS c1 FROM test WHERE __key = 1"));
+
+        assertThrows(HazelcastSqlException.class, () -> query("SELECT JSON_QUERY(this, '$' ERROR ON ERROR) AS c1 FROM test WHERE __key = 2"));
+        assertNull(querySingleValue("SELECT JSON_QUERY(this, '$' NULL ON ERROR) AS c1 FROM test WHERE __key = 2"));
+        assertEquals(new HazelcastJsonValue("[]"),
+                querySingleValue("SELECT JSON_QUERY(this, '$' EMPTY ARRAY ON ERROR) AS c1 FROM test WHERE __key = 2"));
+        assertEquals(new HazelcastJsonValue("{}"),
+                querySingleValue("SELECT JSON_QUERY(this, '$' EMPTY OBJECT ON ERROR) AS c1 FROM test WHERE __key = 2"));
+    }
+
+    @Test
+    public void when_defaultWrapperBehaviorIsSpecified_queryWorks() {
+        initComplexObject();
+        execute("CREATE MAPPING test TYPE IMap OPTIONS ('keyFormat'='bigint', 'valueFormat'='json')");
+        assertEquals("[1,\"2\",3,{\"t\":1}]",
+                querySingleValue("SELECT JSON_QUERY(this, '$[0]') FROM test").toString());
+        assertEquals("{\"t\":1}",
+                querySingleValue("SELECT JSON_QUERY(this, '$[1]') FROM test").toString());
+        assertThrows(HazelcastSqlException.class,
+                () -> querySingleValue("SELECT JSON_QUERY(this, '$[2]' ERROR ON ERROR) FROM test"));
+    }
+
+    @Test
+    public void when_noArrayWrapperSpecified_queryWorks() {
+        initComplexObject();
+        execute("CREATE MAPPING test TYPE IMap OPTIONS ('keyFormat'='bigint', 'valueFormat'='json')");
+
+        assertEquals("[1,\"2\",3,{\"t\":1}]",
+                querySingleValue("SELECT JSON_QUERY(this, '$[0]' WITHOUT ARRAY WRAPPER) FROM test")
+                        .toString());
+        assertEquals("{\"t\":1}",
+                querySingleValue("SELECT JSON_QUERY(this, '$[1]' WITHOUT ARRAY WRAPPER) FROM test")
+                        .toString());
+        assertThrows(HazelcastSqlException.class,
+                () -> querySingleValue("SELECT JSON_QUERY(this, '$[2]' WITHOUT ARRAY WRAPPER ERROR ON ERROR) "
+                        + "FROM test"));
+    }
+
+    @Test
+    public void when_conditionalArrayWrapperSpecified_queryWorks() {
+        initComplexObject();
+        execute("CREATE MAPPING test TYPE IMap OPTIONS ('keyFormat'='bigint', 'valueFormat'='json')");
+
+        assertEquals("[1,\"2\",3,{\"t\":1}]",
+                querySingleValue("SELECT JSON_QUERY(this, '$[0]' WITH CONDITIONAL ARRAY WRAPPER) FROM test")
+                        .toString());
+        assertEquals("{\"t\":1}",
+                querySingleValue("SELECT JSON_QUERY(this, '$[1]' WITH CONDITIONAL ARRAY WRAPPER) FROM test")
+                        .toString());
+        assertEquals("[3]",
```

Comment:
This should actually be equal to `3`, not to `[3]`, if I'm not mistaken. "Conditional" (i guess) means to wrap, if there's more than 1 element.

Reply:
That would be true for `JSON_VALUE`, but `JSON_QUERY` is actually supposed to throw an error if the result is a primitive (and provided Wrapper is not specified). They're mutually exclusive in their behavior in that sense, JSON_QUERY is for objects/arrays only and JSON_VALUE is for primitives only. 

Fwiw I think we should keep these functions close to spec (though not everything is implemented in this PR, namely output clause for JSON_QUERY is not in this PR) and provide flexible behavior like that as part of the json operator.

Reply:
ok, got it. I was wrong.